api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -140,6 +141,31 @@ public StructType partitionType() {
     return lazyPartitionType;
   }
 
+  public StructType originalPartitionType() {
+    if (schema.idsToOriginal().size() == 0) {
+      return partitionType();
+    }
+    if (lazyOriginalPartitionType == null) {
+      synchronized (this) {
+        if (lazyOriginalPartitionType == null) {
+          List<Types.NestedField> structFields = Lists.newArrayListWithExpectedSize(fields.length);
+
+          for (PartitionField field : fields) {
+            Type sourceType = schema.findType(field.sourceId());
```

Comment:
Do we not have a copy constructor here? Seems like we are doing a lot of lookups here for just reassigning the id.

Reply:
Looking into the code we don't have this, but it may be nice to just add a public withId(id) method? like the "asOptional" and "asRequired" methods